File: Backend/src/main/java/lan/dk/podcastserver/manager/worker/updater/RSSUpdater.java
package lan.dk.podcastserver.manager.worker.updater;

import io.vavr.collection.HashSet;
import io.vavr.collection.Set;
import lan.dk.podcastserver.entity.Cover;
import lan.dk.podcastserver.entity.Item;
import lan.dk.podcastserver.entity.Podcast;
import lan.dk.podcastserver.service.ImageService;
import lan.dk.podcastserver.service.JdomService;
import lan.dk.podcastserver.service.SignatureService;
import lan.dk.podcastserver.service.properties.PodcastServerParameters;
import lombok.extern.slf4j.Slf4j;
import org.apache.commons.lang3.StringUtils;
import org.jdom2.Element;
import org.jdom2.Namespace;
import org.springframework.stereotype.Component;

import javax.validation.Validator;
import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;

import static java.util.Objects.isNull;
import static java.util.Objects.nonNull;

@Slf4j
@Component("RSSUpdater")
public class RSSUpdater extends AbstractUpdater {

    private static final Namespace MEDIA = Namespace.getNamespace("media", "http://search.yahoo.com/mrss/");
    private static final Namespace FEED_BURNER = Namespace.getNamespace("feedburner", "http://rssnamespace.org/feedburner/ext/1.0");

    private final JdomService jdomService;
    private final ImageService imageService;

    public RSSUpdater(PodcastServerParameters podcastServerParameters, SignatureService signatureService, Validator validator, JdomService jdomService, ImageService imageService) {
        super(podcastServerParameters, signatureService, validator);
        this.jdomService = jdomService;
        this.imageService = imageService;
    }


    public Set<Item> getItems(Podcast podcast) {
        return jdomService
                .parse(podcast.getUrl())
                .map(p -> p.getRootElement().getChild("channel").getChildren("item"))
                .map(HashSet::ofAll)
                .map(this::elementsToItems)
                .getOrElse(HashSet::empty);
    }

    private Set<Item> elementsToItems(Set<Element> elements) {
        return HashSet.ofAll(elements)
                    .filter(this::hasEnclosure)
                    .map(this::extractItem);
    }

    private Boolean hasEnclosure(Element item) {
        return item.getChild("enclosure") != null || item.getChild("origEnclosureLink", FEED_BURNER) != null;
    }

    private Item extractItem(Element item) {
        // Gestion des cas pour l'url :
        return Item.builder()
                    .title(item.getChildText("title"))
                    .pubDate(getPubDate(item))
                    .description(item.getChildText("description"))
                    .mimeType(item.getChild("enclosure").getAttributeValue("type"))
                    .length(lengthOf(item))
                    .cover(coverOf(item))
                    .url(urlOf(item))
                .build();
    }

    private long lengthOf(Element item) {
        return (StringUtils.isNotEmpty(item.getChild("enclosure").getAttributeValue("length")))
                ? Long.parseLong(item.getChild("enclosure").getAttributeValue("length"))
                : 0L;
    }

    private String urlOf(Element element) {
        if (nonNull(element.getChild("origEnclosureLink", FEED_BURNER))) {
            return element.getChildText("origEnclosureLink", FEED_BURNER);
        }

        return element.getChild("enclosure").getAttributeValue("url");
    }

    private Cover coverOf(Element element) {
        Element thumbnail = element.getChild("thumbnail", MEDIA);
        if (isNull(thumbnail)) {
            return null;
        }

        if (nonNull(thumbnail.getAttributeValue("url"))) {
            return imageService.getCoverFromURL(thumbnail.getAttributeValue("url"));
        }

        return imageService.getCoverFromURL(thumbnail.getText());
    }

    private ZonedDateTime getPubDate(Element item) {
        try {
            return ZonedDateTime.parse(item.getChildText("pubDate").replace(" PST", " +0800"), DateTimeFormatter.RFC_1123_DATE_TIME);
        } catch (Exception e) {
            log.error("Problem during date parsing of {}", item.getChildText("title"), e);
            // No better idea than returning null for unparseable date
            return null;
        }
    }

    @Override
    public String signatureOf(Podcast podcast) {
        return signatureService.generateSignatureFromURL(podcast.getUrl());
    }

    @Override
    public Type type() {
        return new Type("RSS", "RSS");
    }

    @Override
    public Integer compatibility(String url) {
        return Integer.MAX_VALUE-1;
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/update/updaters/rss/RSSUpdater.kt
package com.github.davinkevin.podcastserver.update.updaters.rss

import com.github.davinkevin.podcastserver.service.image.ImageService
import com.github.davinkevin.podcastserver.update.fetchCoverUpdateInformation
import com.github.davinkevin.podcastserver.update.updaters.ItemFromUpdate
import com.github.davinkevin.podcastserver.update.updaters.PodcastToUpdate
import com.github.davinkevin.podcastserver.update.updaters.Type
import com.github.davinkevin.podcastserver.update.updaters.Updater
import org.jdom2.Element
import org.jdom2.Namespace
import org.jdom2.input.SAXBuilder
import org.slf4j.LoggerFactory
import org.springframework.core.io.ByteArrayResource
import org.springframework.util.DigestUtils
import org.springframework.web.client.RestClient
import org.springframework.web.client.body
import java.net.URI
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

private val MEDIA = Namespace.getNamespace("media", "http://search.yahoo.com/mrss/")
private val FEED_BURNER = Namespace.getNamespace("feedburner", "http://rssnamespace.org/feedburner/ext/1.0")
private val ITUNES_NS = Namespace.getNamespace("itunes", "http://www.itunes.com/dtds/podcast-1.0.dtd")!!

class RSSUpdater(
    private val imageService: ImageService,
    private val rcb: RestClient.Builder
) : Updater {

    private val log = LoggerFactory.getLogger(RSSUpdater::class.java)

    override fun findItems(podcast: PodcastToUpdate): List<ItemFromUpdate> {
        val resource = fetchRss(podcast.url)
            ?: return emptyList()

        val xml = SAXBuilder().build(resource.inputStream)

        val channel = xml.rootElement.getChild("channel")
            ?: return emptyList()

        val rssCover = channel.getChild("image")?.getChildText("url")
        val itunesCover = channel.getChild("image", ITUNES_NS)?.getAttributeValue("href")
        val alternativeCoverURL = (rssCover ?: itunesCover)?.let(::URI)

        return channel
            .getChildren("item")
            .toList()
            .filter(::hasEnclosure)
            .map { findItem(it, alternativeCoverURL) }
    }

    private fun findItem(elem: Element, alternativeCoverURL: URI?): ItemFromUpdate {
        val coverUrl = findCover(elem) ?: alternativeCoverURL
        val cover = coverUrl?.let(imageService::fetchCoverUpdateInformation)

        return ItemFromUpdate(
            title = elem.getChildText("title"),
            pubDate = getPubDate(elem),
            description = elem.getChildText("description"),
            cover = cover,
            url = urlOf(elem),
            guid = elem.getChildText("guid"),

            length = elem.enclosure().getAttributeValue("length")?.toLong(),
            mimeType = mimeTypeOf(elem)
        )
    }

    private fun findCover(elem: Element): URI? {
        val thumbnail = elem.getChild("thumbnail", MEDIA)
            ?: return null

        return URI.create(thumbnail.getAttributeValue("url") ?: thumbnail.text)
    }

    private fun hasEnclosure(item: Element) = item.enclosure() != null

    private fun urlOf(element: Element): URI {
        val url = if (element.getChild("origEnclosureLink", FEED_BURNER) != null)
            element.getChildText("origEnclosureLink", FEED_BURNER)
        else
            element.enclosure().getAttributeValue("url")

        return URI(url.replace(" ", "+"))
    }

    private fun mimeTypeOf(elem: Element): String {
        val type = elem.enclosure().getAttributeValue("type")
        return if (!type.isNullOrEmpty()) type else "unknown/unknown"
    }

    private fun getPubDate(item: Element): ZonedDateTime {
        val pubDate = item.getChildText("pubDate") ?: ""
        val date = when {
            "EDT" in pubDate -> pubDate.replace("EDT", "+0600")
            "PST" in pubDate -> pubDate.replace("PST", "+0800")
            "PDT" in pubDate -> pubDate.replace("PDT", "+0900")
            else -> pubDate
        }

        return Result.runCatching {
            ZonedDateTime.parse(date, DateTimeFormatter.RFC_1123_DATE_TIME)
        } .onFailure {
            log.error("Problem during date parsing of \"{}\" caused by {}", item.getChildText("title"), it.message)
        }
            .getOrDefault(ZonedDateTime.now())
    }

    override fun signatureOf(url: URI): String {
        val resource = fetchRss(url)
            ?: return ""

        return DigestUtils.md5DigestAsHex(resource.inputStream)
    }

    private fun fetchRss(url: URI): ByteArrayResource? {
        return runCatching {
            rcb.clone()
                .baseUrl(url.toASCIIString())
                .build()
                .get()
                .retrieve()
                .body<ByteArrayResource>()
        }
            .getOrNull()
    }

    override fun type() = Type("RSS", "RSS")

    override fun compatibility(url: String): Int = when {
        url.startsWith("http", true) -> Integer.MAX_VALUE - 1
        else -> Integer.MAX_VALUE
    }
}

private fun Element.enclosure() = this.getChild("enclosure")